### DISCUSSION TITLE
How to get a tar.gz file from UploadFile

### DISCUSSION THREAD

**Discussion by rujutashinde** [fastapi/fastapi#6327] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.put('/results', response_model=None)
def put_results(config_file: UploadFile = File(...), payload_file: UploadFile = File(...))
	content = payload_file.file.read()
	file = tarfile.open(content, "r:gz")

@app.put('/results', response_model=None)
def put_results(config_file: UploadFile = File(...), payload_file: UploadFile = File(...))
	content = payload_file.file.read()
	file = tarfile.open(content, "r:gz")
```

### Description

I am getting a tar.gz file from UploadFile. I want to extract this tarfile and process it further.
I tried reading the tar.gz file but it gives error, using the uploaded file also gives an error 

Below is what I tried

```
@app.put('/results', response_model=None)
def put_results(config_file: UploadFile = File(...), payload_file: UploadFile = File(...))
	file = tarfile.open(payload_file.file, "r:gz")
	log.debug(file.getnames())

```
ERROR:
```

Traceback (most recent call last):
  File "/home/rushinde/teflo_changes/drouter_work/droute-api/./drouteapi/libs/validation.py", line 83, in validate_config_file
    file = tarfile.open(payload_file.file)
  File "/home/rushinde/Envs/py39/lib/python3.9/tarfile.py", line 1611, in open
    return func(name, "r", fileobj, **kwargs)
  File "/home/rushinde/Envs/py39/lib/python3.9/tarfile.py", line 1675, in gzopen
    fileobj = GzipFile(name, mode + "b", compresslevel, fileobj)
  File "/home/rushinde/.pyenv/versions/3.9.9/lib/python3.9/gzip.py", line 173, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
TypeError: expected str, bytes or os.PathLike object, not SpooledTemporaryFile
```

```
@app.put('/results', response_model=None)
def put_results(config_file: UploadFile = File(...), payload_file: UploadFile = File(...))
	content = payload_file.file.read()
	file = tarfile.open(content, "r:gz")
```
ERROR:

```
Traceback (most recent call last):
  File "/home/rushinde/teflo_changes/drouter_work/droute-api/./drouteapi/libs/validation.py", line 81, in validate_config_file
    file = tarfile.open(content)
  File "/home/rushinde/Envs/py39/lib/python3.9/tarfile.py", line 1611, in open
    return func(name, "r", fileobj, **kwargs)
  File "/home/rushinde/Envs/py39/lib/python3.9/tarfile.py", line 1675, in gzopen
    fileobj = GzipFile(name, mode + "b", compresslevel, fileobj)
  File "/home/rushinde/.pyenv/versions/3.9.9/lib/python3.9/gzip.py", line 173, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
ValueError: embedded null byte

```
This is how i am using the curl command

`curl -X 'PUT'   'http://127.0.0.1:8000/results'   -H 'accept: application/json'   -H 'Content-Type: multipart/form-data'   -F 'config_file=@user_input.json;type=application/json'   -F 'payload_file=@payload.tar.gz;type=application/gzip' `

Thank you !

### Operating System

Linux

### FastAPI Version

fastapi==0.70.0

### Python Version

3.9

**Comment by jgould22** [comment, marked as answer]
The docs for that function are https://docs.python.org/3/library/tarfile.html#tarfile.open

You have a file like object as it says in your first error (`TypeError: expected str, bytes or os.PathLike object, not SpooledTemporaryFile`), however you are passing it in as the first parameter which is the `name` parameter. With your file object you will want to use the `fileobj` parameter. Try and change your open command to something like 
```
	file = tarfile.open(fileobj=payload_file.file, mode="r:gz")
```

**Comment by rujutashinde** [comment]
Thanks @jgould22  That helped. I had another workaround below which I used instead, but the solution you provided is what I was looking for

```
from pathlib import Path
from tempfile import NamedTemporaryFile
  try:
      suffix = Path(payload_file.filename).suffix
      with NamedTemporaryFile(delete=False, suffix=suffix) as tmp:
          shutil.copyfileobj(payload_file.file, tmp)
          self.tmp_path = Path(tmp.name)
  finally:
      payload_file.file.close()
```
